src/protocol/context/mod.rs

```diff
@@ -0,0 +1,94 @@
+use crate::ff::Field;
+use crate::helpers::messaging::{Gateway, Mesh};
+use crate::helpers::Role;
+use crate::protocol::mul::SecureMul;
+use crate::protocol::prss::{
+    Endpoint as PrssEndpoint, IndexedSharedRandomness, SequentialSharedRandomness,
+};
+use crate::protocol::reveal::Reveal;
+use crate::protocol::{Step, Substep};
+use crate::secret_sharing::SecretSharing;
+use std::sync::Arc;
+
+mod malicious;
+mod semi_honest;
+
+pub use malicious::MaliciousProtocolContext;
+pub use semi_honest::SemiHonestProtocolContext;
+
+/// Context used by each helper to perform secure computation. Provides access to shared randomness
+/// generator and communication channel.
+pub trait ProtocolContext<F: Field>:
```

Comment:
As below, it might be possible - based on what I'm seeing anyway - to remove the F generic parameter from this and attach it to the Share type.

Reply:
I tried to look into that and I ran into some issues

* It does not seem possible to constraint the `Share` type 

this does not work

```rust
pub trait ProtocolContext:
    Clone
    + SecureMul<Share<F> = <Self as ProtocolContext>::Share<F>>
    + Reveal<Share<F> = <Self as ProtocolContext>::Share<F>>
```

and removing `F` does not work either:

```rust
pub trait ProtocolContext:
    Clone
    + SecureMul<Share = <Self as ProtocolContext>::Share>
    + Reveal<Share = <Self as ProtocolContext>::Share>
```

I also don't like that `MaliciousContext` needs to be generic over `F` to store accumulator but moving `F` to `Share` will force it to support multiplication of any `G: Field` which is wrong.

Reply:
Ah, that's right.  I had forgotten that reveal was weird.